kubernetes-deployment/maintainlib/etcdfix.py

```diff
@@ -0,0 +1,236 @@
+            return
+
+        commandline = "tar -xvf {0}.tar && sudo ./{0}/{1}.sh".format("etcd-reconfiguration-restart", "restart-etcd-server")
+
+        if common.ssh_shell_paramiko(bad_node_config, commandline) == False:
+            return
+
+        self.logger.info("Successfully restarting bad etcd server on node {0}".format(bad_node_config["nodename"]))
+
+        if self.clean_flag:
+            self.delete_packege(bad_node_config)
+
+
+
+    def get_etcd_leader_node(self):
+
+        # Get leader node.
+        host_list = list()
+
+        for host in self.cluster_config['mastermachinelist']:
+            host_list.append((self.cluster_config['mastermachinelist'][host]['hostip'], 4001))
+
+        client = etcd.Client(host=tuple(host_list), allow_reconnect=True)
+
+        etcdid = client.leader['name']
+
+        for host in self.cluster_config['mastermachinelist']:
+            if etcdid == self.cluster_config['mastermachinelist'][host]['etcdid']:
+                self.logger.debug("Current leader of etcd-cluster: {0}".format(self.cluster_config['mastermachinelist'][host]))
+                return self.cluster_config['mastermachinelist'][host]
+
+        self.logger.error("Can't find the leader of etcd.")
+        return None
+
+
+
+    def get_etcd_peer_ip_list(self, bad_node_config):
+
+        etcd_cluster_ips_peer = ""
+        separated = ""
+
+        host_list = list()
+
+        for host in self.cluster_config['mastermachinelist']:
+            host_list.append((self.cluster_config['mastermachinelist'][host]['hostip'], 4001))
+
+        client = etcd.Client(host=tuple(host_list), allow_reconnect=True)
+
+        member_dict = client.members
+        for member_hash in member_dict:
+
+            etcd_id = member_dict[member_hash]['name']
+            peer_url = member_dict[member_hash]['peerURLs'][0]
+
+            if etcd_id == "":
+                # new member before announcing, etcdid will be empty.
+                continue
+
+            ip_peer = "{0}={1}".format(etcd_id, peer_url)
+
+            etcd_cluster_ips_peer = etcd_cluster_ips_peer + separated + ip_peer
+
+            separated = ","
+
+        new_etcd_id = bad_node_config['etcdid']
+        peer_url = bad_node_config['hostip']
+        ip_peer = "{0}=http://{1}:2380".format(new_etcd_id, peer_url)
+        etcd_cluster_ips_peer = etcd_cluster_ips_peer + separated + ip_peer
+
+        self.logger.debug("New etcd-initial-cluster: {0}".format(etcd_cluster_ips_peer))
+
+        return etcd_cluster_ips_peer
+
+
+
+    def run(self):
+
+        self.logger.info("Begin to fix etcd-cluster's bad member!")
+
+        bad_node_config = self.bad_node_config
+
+        self.logger.debug("Bad node information: {0}".format(str(bad_node_config)))
+
+        self.stop_bad_etcd_server(bad_node_config)
+
+        while True:
+
+            good_node_config = self.get_etcd_leader_node()
+
+            if good_node_config['nodename'] != self.bad_node_config['nodename']:
```

Comment:
>bad_node_config [](start = 52, length = 15)

what if bad node and good node don't overlap? you should add a unit test

Reply:
I have no idea about what to test. Can you give me an example?

Reply:
I see. I suggest to add a comment: "waiting for the bad node to demote from leader"

---
In reply to: [172111494](https://github.com/Microsoft/pai/pull/345#discussion_r172111494) [](ancestors = 172111494)

Reply:
Done